[AIG]Websockets Docs (#20898)
cloudflare/cloudflare-docs: src/content/docs/ai-gateway/configuration/websockets-api/index.mdx

---
title: WebSockets API
pcx_content_type: configuration
sidebar:
  group:
    badge: Beta
---

The AI Gateway WebSockets API provides a persistent connection for AI interactions, eliminating repeated handshakes and reducing latency. This API is divided into two categories:

> 1. **Non-Realtime APIs** - Supports standard WebSocket communication for AI providers, including those that do not natively support WebSockets.

Reviewer: ```suggestion
- **Non-Realtime APIs** - Supports standard WebSocket communication for AI providers, including those that do not natively support WebSockets.
```

> 2. **Realtime APIs** - Designed for AI providers that offer low-latency, multimodal interactions over WebSockets.

Reviewer: ```suggestion
- **Realtime APIs** - Designed for AI providers that offer low-latency, multimodal interactions over WebSockets.
```

## When to use WebSockets?

WebSockets are long-lived TCP connections that enable bi-directional, real-time and non realtime communication between client and server. Unlike HTTP connections, which require repeated handshakes for each request, WebSockets maintain the connection, supporting continuous data exchange with reduced overhead. WebSockets are ideal for applications needing low-latency, real-time data, such as voice assistants.

## Key benefits

> - **Reduced Overhead**: Avoid overhead of repeated handshakes and TLS negotiations by maintaining a single, persistent connection.

Reviewer: ```suggestion
- **Reduced overhead**: Avoid overhead of repeated handshakes and TLS negotiations by maintaining a single, persistent connection.
```

> - **Provider Compatibility**: Works with all AI providers in AI Gateway. Even if your chosen provider does not support WebSockets, we handle it for you, managing the requests to your preferred AI provider.

Reviewer: ```suggestion
- **Provider compatibility**: Works with all AI providers in AI Gateway. Even if your chosen provider does not support WebSockets, Cloudflare handles it for you, managing the requests to your preferred AI provider.
```

| **Purpose**             | Supports WebSocket-based AI interactions with providers that do not natively support WebSockets. | Enables real-time, multimodal AI interactions for providers that offer dedicated WebSocket endpoints.                                 |
| **Use Case**            | Text-based queries and responses, such as LLM requests.                                          | Streaming responses for voice, video, and live interactions.                                                                          |
| **AI Provider Support** | [All AI providers in AI Gateway.](/ai-gateway/providers)                                         | [Limited to providers offering real-time WebSocket APIs.](/ai-gateway/configuration/websockets-api/realtime-api/#supported-providers) |
| **Streaming Support**   | AI Gateway handles streaming via WebSockets.                                                     | Providers natively support real-time data streaming.                                                                                  |

> For details on implementation, see the next section:

Reviewer: ```suggestion
For details on implementation, refer to the next sections:
```
